File: uploadservice/src/main/java/net/gotev/uploadservice/UploadTask.java
package net.gotev.uploadservice;

import android.content.Intent;

import net.gotev.uploadservice.data.UploadFile;
import net.gotev.uploadservice.data.UploadInfo;
import net.gotev.uploadservice.data.UploadTaskParameters;
import net.gotev.uploadservice.logger.UploadServiceLogger;
import net.gotev.uploadservice.network.ServerResponse;
import net.gotev.uploadservice.tasklistener.BroadcastHandler;
import net.gotev.uploadservice.tasklistener.NotificationHandler;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Date;
import java.util.Iterator;
import java.util.List;

/**
 * Base class to subclass when creating upload tasks. It contains the logic common to all the tasks,
 * such as notification management, status broadcast, retry logic and some utility methods.
 *
 * @author Aleksandar Gotev
 */
public abstract class UploadTask implements Runnable {

    private static final String LOG_TAG = UploadTask.class.getSimpleName();

    /**
     * Reference to the upload service instance.
     */
    protected UploadService service;

    /**
     * Contains all the parameters set in {@link UploadRequest}.
     */
    protected UploadTaskParameters params = null;

    /**
     * Contains the absolute local path of the successfully uploaded files.
     */
    private final ArrayList<UploadFile> successfullyUploadedFiles = new ArrayList<>();

    /**
     * Flag indicating if the operation should continue or is cancelled. You should never
     * explicitly set this value in your subclasses, as it's written by the Upload Service
     * when you call {@link UploadService#stopUpload(String)}. If this value is false, you should
     * terminate your upload task as soon as possible, so be sure to check the status when
     * performing long running operations such as data transfer. As a rule of thumb, check this
     * value at every step of the upload protocol you are implementing, and after that each chunk
     * of data that has been successfully transferred.
     */
    protected boolean shouldContinue = true;

    private long lastProgressNotificationTime = 0;
    private BroadcastHandler broadcastHandler;
    private NotificationHandler notificationHandler;

    /**
     * Total bytes to transfer. You should initialize this value in the
     * {@link UploadTask#upload()} method of your subclasses, before starting the upload data
     * transfer.
     */
    protected long totalBytes;

    /**
     * Total transferred bytes. You should update this value in your subclasses when you upload
     * some data, and before calling {@link UploadTask#broadcastProgress(long, long)}
     */
    protected long uploadedBytes;

    /**
     * Start timestamp of this upload task.
     */
    private final long startTime;

    /**
     * Counter of the upload attempts that has been made;
     */
    private int attempts;

    /**
     * Implementation of the upload logic.
     *
     * @throws Exception if an error occurs
     */
    abstract protected void upload() throws Exception;

    /**
     * Implement in subclasses to be able to do something when the upload is successful.
     */
    protected void onSuccessfulUpload() {
    }

    public UploadTask() {
        startTime = new Date().getTime();
    }

    /**
     * Initializes the {@link UploadTask}.<br>
     * Override this method in subclasses to perform custom task initialization and to get the
     * custom parameters set in {@link UploadRequest#initializeIntent(Intent)} method.
     *
     * @param service Upload Service instance. You should use this reference as your context.
     * @param intent  intent sent to the service to start the upload
     * @throws IOException if an I/O exception occurs while initializing
     */
    protected void init(UploadService service, int notificationID, Intent intent) throws IOException {
        this.params = intent.getParcelableExtra(UploadService.PARAM_TASK_PARAMETERS);
        this.service = service;
        broadcastHandler = new BroadcastHandler(service, params.getId());

        UploadNotificationConfig config = params.getNotificationConfig();
        if (config != null) {
            notificationHandler = new NotificationHandler(service, notificationID, params.getId(), config);
        } else {
            notificationHandler = null;
        }
    }

    @Override
    public final void run() {

        if (notificationHandler != null) {
            notificationHandler.initialize(new UploadInfo(params.getId()));
        }

        attempts = 0;

        long errorDelay = UploadServiceConfig.INSTANCE.getRetryPolicy().getInitialWaitTimeSeconds();

        while (attempts <= params.getMaxRetries() && shouldContinue) {
            attempts++;

            uploadedBytes = 0;

            try {
                upload();
                break;

            } catch (Exception exc) {
                if (!shouldContinue) {
                    break;
                } else if (attempts > params.getMaxRetries()) {
                    broadcastError(exc);
                } else {
                    UploadServiceLogger.INSTANCE.error(LOG_TAG, "Error in uploadId " + params.getId()
                            + " on attempt " + attempts
                            + ". Waiting " + errorDelay / 1000 + "s before next attempt. ", exc);

                    long beforeSleepTs = System.currentTimeMillis();

                    while (shouldContinue && System.currentTimeMillis() < (beforeSleepTs + errorDelay * 1000)) {
                        try {
                            Thread.sleep(2000);
                        } catch (Throwable ignored) {
                        }
                    }

                    errorDelay *= UploadServiceConfig.INSTANCE.getRetryPolicy().getMultiplier();
                    if (errorDelay > UploadServiceConfig.INSTANCE.getRetryPolicy().getMaxWaitTimeSeconds()) {
                        errorDelay = UploadServiceConfig.INSTANCE.getRetryPolicy().getMaxWaitTimeSeconds();
                    }
                }
            }
        }

        if (!shouldContinue) {
            broadcastCancelled();
        }
    }

    private boolean shouldThrottle(final long uploadedBytes, final long totalBytes) {
        long currentTime = System.currentTimeMillis();

        if (uploadedBytes < totalBytes && currentTime < lastProgressNotificationTime + UploadServiceConfig.INSTANCE.getUploadProgressNotificationIntervalMillis()) {
            return true;
        }

        lastProgressNotificationTime = currentTime;
        return false;
    }

    /**
     * Broadcasts a progress update.
     *
     * @param uploadedBytes number of bytes which has been uploaded to the server
     * @param totalBytes    total bytes of the request
     */
    protected final void broadcastProgress(final long uploadedBytes, final long totalBytes) {

        if (shouldThrottle(uploadedBytes, totalBytes)) return;

        UploadServiceLogger.INSTANCE.debug(LOG_TAG, "Broadcasting upload progress for " + params.getId()
                + ": " + uploadedBytes + " bytes of " + totalBytes);

        final UploadInfo uploadInfo = getUploadInfo();

        broadcastHandler.onProgress(uploadInfo);

        if (notificationHandler != null) {
            notificationHandler.onProgress(uploadInfo);
        }
    }

    /**
     * Broadcasts a completion status update and informs the {@link UploadService} that the task
     * executes successfully.
     * Call this when the task has completed the upload request and has received the response
     * from the server.
     *
     * @param response response got from the server
     */
    protected final void broadcastCompleted(final ServerResponse response) {

        UploadServiceLogger.INSTANCE.debug(LOG_TAG, "Broadcasting upload " + (response.isSuccessful() ? "completed" : "error")
                + " for " + params.getId());

        if (response.isSuccessful()) {
            onSuccessfulUpload();

            if (params.getAutoDeleteSuccessfullyUploadedFiles() && !successfullyUploadedFiles.isEmpty()) {
                for (UploadFile file : successfullyUploadedFiles) {
                    if (file.getHandler().delete(service)) {
                        UploadServiceLogger.INSTANCE.info(LOG_TAG, "Successfully deleted: "
                                + file.getPath());
                    } else {
                        UploadServiceLogger.INSTANCE.error(LOG_TAG, "Unable to delete: "
                                + file.getPath());
                    }
                }
            }
        }

        final UploadInfo uploadInfo = getUploadInfo();

        broadcastHandler.onCompleted(uploadInfo, response);

        if (notificationHandler != null) {
            notificationHandler.onCompleted(uploadInfo, response);
        }

        service.taskCompleted(params.getId());
    }

    /**
     * Broadcast a cancelled status.
     * This called automatically by {@link UploadTask} when the user cancels the request,
     * and the specific implementation of {@link UploadTask#upload()} either
     * returns or throws an exception. You should never call this method explicitly in your
     * implementation.
     */
    protected final void broadcastCancelled() {

        UploadServiceLogger.INSTANCE.debug(LOG_TAG, "Broadcasting cancellation for upload with ID: " + params.getId());

        final UploadInfo uploadInfo = getUploadInfo();

        broadcastHandler.onCancelled(uploadInfo);

        if (notificationHandler != null) {
            notificationHandler.onCancelled(uploadInfo);
        }

        service.taskCompleted(params.getId());
    }

    /**
     * Broadcasts an error.
     * This called automatically by {@link UploadTask} when the specific implementation of
     * {@link UploadTask#upload()} throws an exception and there aren't any left retries.
     * You should never call this method explicitly in your implementation.
     *
     * @param exception exception to broadcast. It's the one thrown by the specific implementation
     *                  of {@link UploadTask#upload()}
     */
    private void broadcastError(final Exception exception) {

        UploadServiceLogger.INSTANCE.info(LOG_TAG, "Broadcasting error for upload with ID: "
                + params.getId() + ". " + exception.getMessage());

        final UploadInfo uploadInfo = getUploadInfo();

        broadcastHandler.onError(uploadInfo, exception);

        if (notificationHandler != null) {
            notificationHandler.onError(uploadInfo, exception);
        }

        service.taskCompleted(params.getId());
    }

    /**
     * Add a file to the list of the successfully uploaded files and remove it from the file list
     *
     * @param file file on the device
     */
    protected final void addSuccessfullyUploadedFile(UploadFile file) {
        if (!successfullyUploadedFiles.contains(file)) {
            successfullyUploadedFiles.add(file);
            params.getFiles().remove(file);
        }
    }

    /**
     * Adds all the files to the list of successfully uploaded files.
     * This will automatically remove them from the params.getFiles() list.
     */
    protected final void addAllFilesToSuccessfullyUploadedFiles() {
        for (Iterator<UploadFile> iterator = params.getFiles().iterator(); iterator.hasNext(); ) {
            UploadFile file = iterator.next();

            if (!successfullyUploadedFiles.contains(file)) {
                successfullyUploadedFiles.add(file);
            }
            iterator.remove();
        }
    }

    /**
     * Gets the list of all the successfully uploaded files.
     * You must not modify this list in your subclasses! You can only read its contents.
     * If you want to add an element into it,
     * use {@link UploadTask#addSuccessfullyUploadedFile(UploadFile)}
     *
     * @return list of strings
     */
    protected final List<UploadFile> getSuccessfullyUploadedFiles() {
        return successfullyUploadedFiles;
    }

    private UploadInfo getUploadInfo() {
        return new UploadInfo(params.getId(), startTime, uploadedBytes,
                totalBytes, (attempts - 1),
                null,
                successfullyUploadedFiles,
                params.getFiles()
        );
    }

    public final void cancel() {
        this.shouldContinue = false;
    }

}

File: uploadservice/src/main/java/net/gotev/uploadservice/UploadTask.kt
package net.gotev.uploadservice

import android.content.Context
import net.gotev.uploadservice.data.UploadFile
import net.gotev.uploadservice.data.UploadInfo
import net.gotev.uploadservice.data.UploadNotificationConfig
import net.gotev.uploadservice.data.UploadTaskParameters
import net.gotev.uploadservice.exceptions.UploadError
import net.gotev.uploadservice.exceptions.UserCancelledUploadException
import net.gotev.uploadservice.logger.UploadServiceLogger
import net.gotev.uploadservice.network.HttpStack
import net.gotev.uploadservice.network.ServerResponse
import net.gotev.uploadservice.observer.task.UploadTaskObserver
import java.io.IOException
import java.util.ArrayList
import java.util.Date

abstract class UploadTask : Runnable {

    companion object {
        private val TAG = UploadTask::class.java.simpleName
    }

    private var lastProgressNotificationTime: Long = 0

    protected lateinit var context: Context
    lateinit var params: UploadTaskParameters
    lateinit var notificationConfig: UploadNotificationConfig
    var notificationId: Int = 0

    /**
     * Flag indicating if the operation should continue or is cancelled. You should never
     * explicitly set this value in your subclasses, as it's written by the Upload Service
     * when you call [UploadService.stopUpload]. If this value is false, you should
     * terminate your upload task as soon as possible, so be sure to check the status when
     * performing long running operations such as data transfer. As a rule of thumb, check this
     * value at every step of the upload protocol you are implementing, and after that each chunk
     * of data that has been successfully transferred.
     */
    var shouldContinue = true

    private val observers = ArrayList<UploadTaskObserver>(2)

    /**
     * Total bytes to transfer. You should initialize this value in the
     * [UploadTask.upload] method of your subclasses, before starting the upload data
     * transfer.
     */
    var totalBytes: Long = 0

    /**
     * Total transferred bytes. You should update this value in your subclasses when you upload
     * some data, and before calling [UploadTask.onProgress]
     */
    private var uploadedBytes: Long = 0

    /**
     * Start timestamp of this upload task.
     */
    private val startTime: Long = Date().time

    /**
     * Counter of the upload attempts that has been made;
     */
    private var attempts: Int = 0

    private var errorDelay = UploadServiceConfig.retryPolicy.initialWaitTimeSeconds.toLong()

    private val uploadInfo: UploadInfo
        get() = UploadInfo(
            uploadId = params.id,
            startTime = startTime,
            uploadedBytes = uploadedBytes,
            totalBytes = totalBytes,
            numberOfRetries = attempts,
            files = params.files
        )

    /**
     * Implementation of the upload logic.
     *
     * @throws Exception if an error occurs
     */
    @Throws(Exception::class)
    protected abstract fun upload(httpStack: HttpStack)

    private inline fun doForEachObserver(action: UploadTaskObserver.() -> Unit) {
        observers.forEach {
            try {
                action(it)
            } catch (exc: Throwable) {
                UploadServiceLogger.error(TAG, params.id, exc) {
                    "error while dispatching event to observer"
                }
            }
        }
    }

    /**
     * Initializes the [UploadTask].<br></br>
     * Override this method in subclasses to perform custom task initialization and to get the
     * custom parameters set in [UploadRequest.initializeIntent] method.
     *
     * @param context Upload Service instance. You should use this reference as your context.
     * @param intent intent sent to the context to start the upload
     * @throws IOException if an I/O exception occurs while initializing
     */
    @Throws(IOException::class)
    fun init(
        context: Context,
        taskParams: UploadTaskParameters,
        notificationConfig: UploadNotificationConfig,
        notificationId: Int,
        vararg taskObservers: UploadTaskObserver
    ) {
        this.context = context
        this.params = taskParams
        this.notificationId = notificationId
        this.notificationConfig = notificationConfig
        taskObservers.forEach { observers.add(it) }
        performInitialization()
    }

    open fun performInitialization() {}

    private fun resetAttempts() {
        attempts = 0
        errorDelay = UploadServiceConfig.retryPolicy.initialWaitTimeSeconds.toLong()
    }

    /**
     * Executes the upload task and notifies UploadTaskObservers the start of the upload.
     * The body is impelemented on runUpload method.
     */
    override fun run() {
        doForEachObserver {
            onStart(
                uploadInfo,
                notificationId,
                notificationConfig
            )
        }
        resetAttempts()
        runUpload()
    }

    /**
     * Runs the body of upload task. If exception is thrown on the body of upload task it runs the exceptionHandeling method.
     * The exceptionHandeling implementation is moved into it's separate method so that it can be called directly via inherited classes (if needed)
     * In some cases such as S3 where the upload thread happens on a different thread than the upload task
     * and we don't have control over it, we can call the exceptionHandeling manually on observance of an error.
     * This method is called at the end of exceptionHandeling for the number of attempts.
     * Note: "run" resets attempts as well so at the end of exceptionHandeling "run" should not be called. Instead this method is called.
     */
    private fun runUpload() {
        while (attempts <= params.maxRetries && shouldContinue) {
            try {
                resetUploadedBytes()
                upload(UploadServiceConfig.httpStack)
                attempts++
                break
            } catch (exc: Throwable) {
                exceptionHandling(exc)
            }
        }
        if (!shouldContinue) {
            onUserCancelledUpload()
        }
    }

    /**
     * If an exception is thrown while uploading, this method will be executed to do the retry.
     * This method (retry) is automatically called on upload task exception.
     * However, if upload is carried over a separate thread you need to call this manually.
     * You can call this method directly on observance of an error. See S3 implementation for example
     * @param exc: The exception that has been thrown
     */
    fun exceptionHandling(exc: Throwable) {
        if (!shouldContinue) {
            UploadServiceLogger.error(TAG, params.id, exc) { "error while uploading but user requested cancellation." }
        } else if (attempts >= params.maxRetries) {
            onError(exc)
        } else {
            UploadServiceLogger.error(TAG, params.id, exc) { "error on attempt ${attempts + 1}. Waiting ${errorDelay}s before next attempt." }
            val sleepDeadline = System.currentTimeMillis() + errorDelay * 1000
            sleepWhile { shouldContinue && System.currentTimeMillis() < sleepDeadline }
            errorDelay *= UploadServiceConfig.retryPolicy.multiplier.toLong()
            if (errorDelay > UploadServiceConfig.retryPolicy.maxWaitTimeSeconds) {
                errorDelay = UploadServiceConfig.retryPolicy.maxWaitTimeSeconds.toLong()
            }
        }
        attempts++
        runUpload()
    }

    private inline fun sleepWhile(millis: Long = 1000, condition: () -> Boolean) {
        while (condition()) {
            try {
                Thread.sleep(millis)
            } catch (exc: Throwable) {
            }
        }
    }

    protected fun resetUploadedBytes() {
        uploadedBytes = 0
    }

    /**
     * Broadcasts a progress update.
     *
     * @param uploadedBytes number of bytes which has been uploaded to the server
     * @param totalBytes total bytes of the request
     */
    protected fun onProgress(bytesSent: Long) {
        uploadedBytes += bytesSent
        if (shouldThrottle(uploadedBytes, totalBytes)) return
        UploadServiceLogger.debug(TAG, params.id) { "uploaded ${uploadedBytes * 100 / totalBytes}%, $uploadedBytes of $totalBytes bytes" }
        doForEachObserver { onProgress(uploadInfo, notificationId, notificationConfig) }
    }

    /**
     * Broadcasts a completion status update and informs the [UploadService] that the task
     * executes successfully.
     * Call this when the task has completed the upload request and has received the response
     * from the server.
     *
     * @param response response got from the server
     */
    protected fun onResponseReceived(response: ServerResponse) {
        UploadServiceLogger.debug(TAG, params.id) { "upload ${if (response.isSuccessful) "completed" else "error"}" }

        if (response.isSuccessful) {
            if (params.autoDeleteSuccessfullyUploadedFiles) {
                for (file in successfullyUploadedFiles) {
                    if (file.handler.delete(context)) {
                        UploadServiceLogger.info(TAG, params.id) { "successfully deleted: ${file.path}" }
                    } else {
                        UploadServiceLogger.error(TAG, params.id) { "error while deleting: ${file.path}" }
                    }
                }
            }

            doForEachObserver {
                onSuccess(
                    uploadInfo,
                    notificationId,
                    notificationConfig,
                    response
                )
            }
        } else {
            doForEachObserver {
                onError(
                    uploadInfo,
                    notificationId,
                    notificationConfig,
                    UploadError(response)
                )
            }
        }

        doForEachObserver { onCompleted(uploadInfo, notificationId, notificationConfig) }
    }

    /**
     * Broadcast a cancelled status.
     * This called automatically by [UploadTask] when the user cancels the request,
     * and the specific implementation of [UploadTask.upload] either
     * returns or throws an exception. You should never call this method explicitly in your
     * implementation.
     */
    private fun onUserCancelledUpload() {
        UploadServiceLogger.debug(TAG, params.id) { "upload cancelled" }
        onError(UserCancelledUploadException())
    }

    /**
     * Broadcasts an error.
     * This called automatically by [UploadTask] when the specific implementation of
     * [UploadTask.upload] throws an exception and there aren't any left retries.
     * You should never call this method explicitly in your implementation.
     *
     * @param exception exception to broadcast. It's the one thrown by the specific implementation
     * of [UploadTask.upload]
     */
    private fun onError(exception: Throwable) {
        UploadServiceLogger.error(TAG, params.id, exception) { "error" }
        uploadInfo.let {
            doForEachObserver { onError(it, notificationId, notificationConfig, exception) }
            doForEachObserver { onCompleted(it, notificationId, notificationConfig) }
        }
    }

    /**
     * Adds all the files to the list of successfully uploaded files.
     * This will automatically remove them from the params.getFiles() list.
     */
    protected fun setAllFilesHaveBeenSuccessfullyUploaded(value: Boolean = true) {
        params.files.forEach { it.successfullyUploaded = value }
    }

    /**
     * Gets the list of all the successfully uploaded files.
     * You must not modify this list in your subclasses! You can only read its contents.
     * If you want to add an element into it,
     * use [UploadTask.addSuccessfullyUploadedFile]
     *
     * @return list of strings
     */
    protected val successfullyUploadedFiles: List<UploadFile>
        get() = params.files.filter { it.successfullyUploaded }

    fun cancel() {
        shouldContinue = false
    }

    private fun shouldThrottle(uploadedBytes: Long, totalBytes: Long): Boolean {
        val currentTime = System.currentTimeMillis()

        if (uploadedBytes < totalBytes && currentTime < lastProgressNotificationTime + UploadServiceConfig.uploadProgressNotificationIntervalMillis) {
            return true
        }

        lastProgressNotificationTime = currentTime
        return false
    }
}
